starknet-core/src/types/transaction.rs

```diff
@@ -114,6 +118,31 @@ pub struct InvokeFunctionTransaction {
     pub transaction_hash: FieldElement,
     #[serde_as(as = "UfeHex")]
     pub max_fee: FieldElement,
+    #[serde(default)]
+    #[serde_as(as = "Option<UfeHex>")]
+    pub nonce: Option<FieldElement>,
+    #[serde(default)]
+    #[serde_as(as = "Option<UfeHex>")]
+    pub version: Option<FieldElement>,
```

Comment:
Does the `version` field really have to be optional? Looks like it has been retrospectively applied to older blocks as well. Will check to make sure though.

Reply:
We actually had this dicussion before in this thread:

https://github.com/Linear-finance/linear-exchange/pull/27

I think it makes sense for mainnet users to use an older version (a previous commit), especially before we have a fully production-ready mainnet, after which we can start using feature branches for matching testnets.

The thing is attempting to support multiple StarkNet versions inside a single SDK version might not always work, at least not without tradeoffs (in this case, testnet clients are unnecessarily unwrapping `version`).

Reply:
I think you can actually use different versions of the same crate inside a program, which would solve the issue of fetching mainnet and testnet data inside the same application. IMO this use case is a bit niche, but there is this reasonable workaround.

Things should get much better when the gateway API and JSON-RPC spec stablizes more.